### PR TITLE
Build one installer unless we have JRE

### PR DESCRIPTION
**Pull request by M-Davies**
* Builds an installer for the JDK and for JRE if it exists

Closes: #1751 
Signed-off-by: Morgan Davies <morgan.davies@ibm.com>

### CODE REVIEW

**Comment by M-Davies** [comment]
Draft for now while [installer cleanups](https://adoptopenjdk.slack.com/archives/C09NW3L2J/p1590662875341100) are in progress

**Comment by M-Davies** [comment]
Could do with a review when you get time @douph1 :)

**Comment by douph1** [comment]
- I don't know if it the way to go for Linux and Mac installer.
I was asking this just for Windows installer because ( I don't remember why but ) the Windows script can only handle one jre/jdk at a time and building both mean run script two time .. and this is not a good idea for error handling.

- Maybe can you try to avoid so many duplicate by using a function with jre|jdk as param

- I would like you forward this same param to installer job as PRODUCT_CATEGORY param (jre or jdk in lower case)

- Then last but not least, if you can pass another job parameter ARTIFACT with the exact name of the artefact copied it will permit to avoid wildcard things to find wich zip file to unpack.

Thanks

**Comment by M-Davies** [comment]
> Maybe can you try to avoid so many duplicate by using a function with jre|jdk as param
>
If we only need it for windows, then there isn't much point in creating a function for it
> Then last but not least, if you can pass another job parameter ARTIFACT with the exact name of the artefact copied it will permit to avoid wildcard things to find wich zip file to unpack.
> 
Due to parsing issues and the variety of filenames, all we would be doing is moving the parsing to build instead of installer. I think a better solution would be to optimise the `filter` param to only accept the JRE.tar.gz and JDK.tar.gz files. But I'm open to ideas 🙂

**Comment by M-Davies** [comment]
@douph1 @sxa Can I get reviews please? :)

**Comment by douph1** [comment]
@M-Davies can you add PRODUCT_CATEGORY = jdk and PRODUCT_CATEGORY = jre please

**Comment by M-Davies** [comment]
I'd like to test this before merging (somehow). @sxa Could you create a temp pipeline that uses this branch?

File: pipelines/build/common/openjdk_build_pipeline.groovy [hunk 1]
```diff
@@ -309,6 +309,42 @@ class Build {
...
+
+            if (file.contains("-jre")) {
+                
+                context.println("We have a JRE. Running another installer for it...")
+                def jreinstallerJob = context.build job: "build-scripts/release/create_installer_windows",
+                        propagate: true,
+                        parameters: [
+                            context.string(name: 'UPSTREAM_JOB_NUMBER', value: "${env.BUILD_NUMBER}"),
+                            context.string(name: 'UPSTREAM_JOB_NAME', value: "${env.JOB_NAME}"),
+                            context.string(name: 'FILTER', value: "**/OpenJDK*jre_*_windows_*.tar.gz"),
```
[the remark this answers is not in the document]
Comment (M-Davies) [reply pipelines/build/common/openjdk_build_pipeline.groovy:325]: Good spot. I'll update this

**Comment by M-Davies** [comment]
Thanks @douph1 . I'm still awaiting confirmation on a system to test this so please don't merge this yet

**Comment by M-Davies** [comment]
Executing a JDK11 and JDK-Next pipeline to test changes
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk11-pipeline/983/
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk15-pipeline/146/

**Comment by M-Davies** [comment]
Seems the `break` isn't aloud inside for-each groovy loops
```
09:30:16  > Task :compileGroovy FAILED
09:30:16  startup failed:
09:30:16  J:\jenkins\tmp\pipelines\build\common\openjdk_build_pipeline.groovy: 346: the break statement is only allowed inside loops or switches
09:30:16   @ line 346, column 17.
09:30:16                     break
09:30:16                     ^
09:30:16  
09:30:16  1 error
09:30:16  
```

**Comment by M-Davies** [comment]
Resubmitted the pipelines
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk11-pipeline/984/
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk15-pipeline/147/

**Comment by M-Davies** [comment]
https://github.com/AdoptOpenJDK/openjdk-installer/pull/217 is now merged. Kicking off the test pipelines again
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk11-pipeline/985/
- https://ci.adoptopenjdk.net/job/build-scripts/job/openjdk15-pipeline/148/

**Pull request merged**